### PR TITLE
Add support for fill, weight, grade, and optical size to `Icon`

### PR DESCRIPTION
**Pull request by guidezpl**
This PR adds framework support required for Material Symbols ([design doc](http://flutter.dev/go/material-symbols), https://github.com/flutter/flutter/issues/102560). 

> Add 4 members to Icon: fill, weight, grade, and opticalSize. Under the hood, pass these to the engine using the FontVariation API. No guarantees are provided:
> - If used with a font which does not support these axes => no effect
> - No bounds checking for values against the font ⇒ no effect outside variable font's supported range

> Add 4 corresponding members to IconThemeData to customize all icons in a given widget subtree.

Also:
- `debugFillProperties` is completed
- properties are reordered for consistency
- documentation is improved

### Tests
- Basic bounds checking
- Instantiation
- Fallbacks

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/icon.dart [hunk 1]
```diff
@@ -88,33 +94,91 @@ class Icon extends StatelessWidget {
...
   /// [IconButton.iconSize] instead, so that the [IconButton] can make the splash
   /// area the appropriate size as well. The [IconButton] uses an [IconTheme] to
   /// pass down the size to the [Icon].
   final double? size;
 
-  /// The color to use when drawing the icon.
+  /// The fill for drawing the icon.
   ///
-  /// Defaults to the current [IconTheme] color, if any.
+  /// Requires the underlying icon font to support the `FILL` [FontVariation]
```
[the remark this answers is not in the document]
Comment (guidezpl) [reply packages/flutter/lib/src/widgets/icon.dart:107]: Answered elsewhere. I can point out that font file names often include the supported axes

File: packages/flutter/lib/src/cupertino/icon_theme_data.dart [hunk 2]
```diff
@@ -27,18 +27,6 @@ class CupertinoIconThemeData extends IconThemeData with Diagnosticable {
     return resolvedColor == color ? this : copyWith(color: resolvedColor);
   }
 
-  /// Creates a copy of this icon theme but with the given fields replaced with
-  /// the new values.
-  @override
-  CupertinoIconThemeData copyWith({ Color? color, double? opacity, double? size, List<Shadow>? shadows }) {
```
Comment (darrenaustin) [inline packages/flutter/lib/src/cupertino/icon_theme_data.dart:33]: Why are we removing `copyWith`? Don't we still need it?

File: packages/flutter/lib/src/widgets/icon.dart [hunk 3]
```diff
@@ -88,33 +94,95 @@ class Icon extends StatelessWidget {
   ///
   /// Icons occupy a square with width and height equal to size.
   ///
-  /// Defaults to the current [IconTheme] size, if any. If there is no
-  /// [IconTheme], or it does not specify an explicit size, then it defaults to
-  /// 24.0.
+  /// Defaults to the nearest [IconTheme]'s [IconThemeData.size].
```
Comment (darrenaustin) [inline packages/flutter/lib/src/widgets/icon.dart:97]: Do we have a default for when the IconTheme doesn't provide a size?

Same question for all the new properties as well.

**Review by darrenaustin — CHANGES_REQUESTED** [review]
Looks good. Just a question and request for defaults to be documented.

File: packages/flutter/lib/src/widgets/icon.dart (hunk 3 above)
Comment (guidezpl) [reply packages/flutter/lib/src/widgets/icon.dart:97]: There's a fallback, which is documented in IconThemeData.size (and similarly for all properties).

File: packages/flutter/lib/src/cupertino/icon_theme_data.dart (hunk 2 above)
Comment (guidezpl) [reply packages/flutter/lib/src/cupertino/icon_theme_data.dart:33]: Initially thought it was an unnecessary override, but it does return a more specific type, so added it back.

File: packages/flutter/lib/src/widgets/icon.dart (hunk 3 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/widgets/icon.dart:97]: Fair enough. We are somewhat inconsistent on where we document defaults (in the widget or the theme), but we can do a sweep for these later.

**Review by darrenaustin — APPROVED** [review]
LGTM. Thx.

File: packages/flutter/lib/src/widgets/icon.dart (hunk 3 above)
Comment (guidezpl) [reply packages/flutter/lib/src/widgets/icon.dart:97]: I like the idea of documenting defaults only where they're set, in most cases that'll be the widget. In this case, the theme.

File: packages/flutter/lib/src/widgets/icon.dart [hunk 4]
```diff
@@ -88,33 +94,91 @@ class Icon extends StatelessWidget {
...
+
+  /// The grade (granular stroke weight) for drawing the icon.
+  ///
+  /// Requires the underlying icon font to support the `GRAD` [FontVariation]
+  /// axis, otherwise has no effect. Can be negative.
+  ///
+  /// Grade and [weight] both affect a symbol's stroke weight (thickness), but
+  /// grade has a smaller impact on the size of the symbol.
+  ///
+  /// Grade is also available in some text fonts. One can match grade levels
```
[the remark this answers is not in the document]
Comment (davelab6) [reply packages/flutter/lib/src/widgets/icon.dart:146]: https://fonts.google.com/knowledge/introducing_type/introducing_variable_fonts might be a good link

**Pull request merged**